sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -471,7 +537,44 @@ def process_element(self, element):
     # We can either have the _TestStream or the _WatermarkController to emit
     # the elements. We chose to emit in the _WatermarkController so that the
     # element is emitted at the correct watermark value.
-    for event in self.test_stream.events(self.current_index):
+    events = []
+    if self.watermark == MIN_TIMESTAMP:
+      for event in self.test_stream._set_up(self.test_stream.output_tags):
+        events.append(event)
+
+    if self.test_stream_event_channel:
+      try:
+        event = next(self.test_stream_event_channel)
```

Comment:
Don't we already have this translation logic elsewhere. Also, it seems we could unify events-from-memory vs. events-from-grpc more.

Reply:
Ack, unified it with the two methods in the test_stream_impl to return iterators. These will be stored in the global EvaluationContext state.